src/passes/RemoveNonJSOps.cpp

```diff
@@ -14,251 +14,137 @@
 // themselves.
 //
+// You'll find a large wast blob at the end of this module which contains all of
+// the injected intrinsics. We manually copy over any needed intrinsics from
+// this module into the module that we're optimizing after walking the current
+// module.
+//
 
 #include <wasm.h>
 #include <pass.h>
 
 #include "asmjs/shared-constants.h"
 #include "wasm-builder.h"
+#include "wasm-s-parser.h"
 
 namespace wasm {
 
+namespace {
+extern const char *IntrinsicsModule;
+}
+
 struct RemoveNonJSOpsPass : public WalkerPass<PostWalker<RemoveNonJSOpsPass>> {
-  bool needNearestF32 = false;
-  bool needNearestF64 = false;
-  bool needTruncF32 = false;
-  bool needTruncF64 = false;
-  bool needCtzInt32 = false;
-  bool needPopcntInt32 = false;
-  bool needRotLInt32 = false;
-  bool needRotRInt32 = false;
+  std::unique_ptr<Builder> builder;
+  std::unordered_set<Name> neededIntrinsics;
 
   bool isFunctionParallel() override { return false; }
 
   Pass* create() override { return new RemoveNonJSOpsPass; }
 
   void doWalkModule(Module* module) {
+    // Discover all of the intrinsics that we need to inject, lowering all
+    // operations to intrinsic calls while we're at it.
     if (!builder) builder = make_unique<Builder>(*module);
     PostWalker<RemoveNonJSOpsPass>::doWalkModule(module);
 
-    if (needNearestF32) {
-      module->addFunction(createNearest(f32));
-    }
-    if (needNearestF64) {
-      module->addFunction(createNearest(f64));
+    if (neededIntrinsics.size() == 0) {
+      return;
     }
-    if (needTruncF32) {
-      module->addFunction(createTrunc(f32));
-    }
-    if (needTruncF64) {
-      module->addFunction(createTrunc(f64));
-    }
-    if (needCtzInt32) {
-      module->addFunction(createCtz());
-    }
-    if (needPopcntInt32) {
-      module->addFunction(createPopcnt());
-    }
-    if (needRotLInt32) {
-      module->addFunction(createRot(RotLInt32));
-    }
-    if (needRotRInt32) {
-      module->addFunction(createRot(RotRInt32));
+
+    // Parse the wast blob we have at the end of this file.
+    //
+    // TODO: only do this once per invocation of wasm2asm
+    Module intrinsicsModule;
+    std::string input(IntrinsicsModule);
+    SExpressionParser parser(const_cast<char*>(input.c_str()));
+    Element& root = *parser.root;
+    SExpressionWasmBuilder builder(intrinsicsModule, *root[0]);
+
+    std::set<Name> neededFunctions;
+
+    // Iteratively link intrinsics from `intrinsicsModule` into our destination
+    // module, as needed.
+    //
+    // Note that intrinsics often use one another. For example the 64-bit
+    // division intrinsic ends up using the 32-bit ctz intrinsic, but does so
+    // via a native instruction. The loop here is used to continuously reprocess
+    // injected intrinsics to ensure that they never contain non-js ops when
+    // we're done.
+    while (neededIntrinsics.size() > 0) {
+      // TODO: move popcnt to the wast blob below and avoid special handling of
```

Comment:
what is special about popcount?

Reply:
Ah just that the way I wrote it in Rust LLVM would recognize it and emit a popcnt instruction, which is what this part was trying to avoid. I just haven't gotten around to hand-writing a wast blob to remove the need for the manual construction here